### DISCUSSION TITLE
What does `globus endpoint UID not set in .zstash.ini` mean?

### DISCUSSION THREAD

**Discussion by forsyth2** [E3SM-Project/zstash#241] [Questions]
What does `globus endpoint UID not set in .zstash.ini` mean?

**Comment by forsyth2** [comment]
Check https://github.com/E3SM-Project/zstash/blob/main/zstash/globus.py for the correct UUID. Alternatively, check on the Globus website directly.
```
    r"theta.*\.alcf\.anl\.gov": "08925f04-569f-11e7-bef8-22000b9a448b",
    r"blueslogin.*\.lcrc\.anl\.gov": "61f9954c-a4fa-11ea-8f07-0a21f750d19b",
    r"chr.*\.lcrc\.anl\.gov": "61f9954c-a4fa-11ea-8f07-0a21f750d19b",
    r"cori.*\.nersc\.gov": "9d6d99eb-6d04-11e5-ba46-22000b92c6ec",
    r"perlmutter.*\.nersc\.gov": "6bdc7956-fc0f-4ad2-989c-7aa5ee643a79",  # If this doesn't work, use cori
```

Then:
```
$ emacs ~/.zstash.ini
[local]
globus_endpoint_uuid = <paste UUID here>
```
Then, rerun.

**Reply by lukaszlacinski** [reply]
Zstash does not know what Globus endpoint represent the machine, zstash is executed on. Zstash tries to figure it out based on a hostname and hardcoded mapping, https://github.com/E3SM-Project/zstash/blob/main/zstash/globus.py#L21. If the hostname does not maatch any of the regular expressions, it tries to get the local endpoint UUID from ~/..zstash.ini. If there is no Globus endpoint UUID set in the ini file, zstash fails with the error message.
Maybe the error message should be more descriptive, or the .zstash.ini created with a longer comment that explain what the globus_endpoint_uuid is needed for.

**Reply by forsyth2** [reply]
Thanks for the info!! That would explain why Perlmutter causes some issues -- `$hostname` on Perlmutter just begins with `login` and nothing identifying it as Perlmutter. Yeah, a more descriptive error message may be helpful; I can add that (#242).

**Reply by lukaszlacinski** [reply]
Zstash checks a fully qualified domain name of the node, zstash is executed on. I do not have access to Perlmutter. What do you get when you run `hostname --fqdn`?

**Reply by forsyth2** [reply]
```
$ hostname --fqdn
login20
```

**Reply by forsyth2** [reply]
I know @xylar ran into the issue of Perlmutter's `hostname` with `discover_machine` in Mache: https://github.com/E3SM-Project/mache/blob/main/mache/discover.py#L35. Perhaps we could do something similar to that.

**Reply by xylar** [reply]
On Perlmutter you need to use `$NERSC_HOST`.  If it is `unknown`, you need to tell users to edit their `.bashrc.ext` or equivalent to remove the logic that is setting that (a legacy from Hopper and Edison).  There seems to be no other good way to identify that you are on Perlmutter.

It seems good to have a command-line flag to identify the host for zstash but the `~/.zstash.ini` approach is okay as a workaround.

**Reply by forsyth2** [reply]
Good to know, thanks!!

**Comment by vanroekel** [comment]
@lukaszlacinski thanks for the added info.  I'm not sure if I fully understand this yet.  I initiated the zstash on chrysalis and did not see this error.  When I tried to continue on anvil it gave this error.  Moving back to chrysalis allowed it to work again.  `$HOSTNAME` gives blueslogin so seems like it should work like on chrysalis as that variable gives chrlogin.  Any idea why anvil has this error?

**Reply by lukaszlacinski** [reply]
When I log into Anvil (blues.lcrc.anl.gov) and check the hostname, I get:
```
[lukasz@blueslogin1 ~]$ hostname --fqdn
blueslogin1.lcrc.anl.gov`
```
It matches the regex https://github.com/E3SM-Project/zstash/blob/main/zstash/globus.py#L23.
Can you please run `hostname --fqdn` when you log in to Anvil?

**Reply by vanroekel** [reply]
sure no problem -- I get `blueslogin3.lcrc.anl.gov` but I played with this a bit and think I understand the error.  I'm doing the transfer on compute nodes, so if I do the same thing on anvil I get something like `b641.lcrc.anl.gov` which wouldn't match the regex you link.  When I do this on chrysalis I get `chr-XXX.lcrc.*` which still matches the regex in your link.  I think this makes sense now.

**Reply by lukaszlacinski** [reply]
Temporarily, you can set `globus_endpoint.uuid = 61f9954c-a4fa-11ea-8f07-0a21f750d19b` in ~/.zstash.ini. I will add r"b\d+\.lcrc\.anl\.gov" to globus.py to match the Blues compute nodes for the next zstash release.

**Reply by vanroekel** [reply]
thanks for the workaround and for the help understanding this issue!

**Reply by forsyth2** [reply]
> I will add r"b\d+.lcrc.anl.gov" to globus.py to match the Blues compute nodes for the next zstash release.

Created #243 to track that. Thanks!
